### PR TITLE
New opcodes and operations in AVR anal plugin.

### PR DESCRIPTION
**Pull request by killabytenow**
Here follows a pull request for starting a rewriting and refactorization of the AVR anal plugin. Currently is not complete but it establishes the base that I will follow next days/commits.

Anyway is better than previous implementation and now it is executing some little AVR programs.

If you want to test it, doesnt forget to execute the following statements before:

aeim 0x00010000 0x01ff eeprom
aeim 0x00010200 0xffff sram

### CODE REVIEW

File: libr/anal/p/anal_avr.c [hunk 1]
```diff
@@ -17,35 +17,274 @@
 
 #define	AVR_SOFTCAST(x,y) (x+(y*0x100))
 
+typedef struct _cpu_models_tag_ {
+	char *model;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: this is const

File: libr/anal/p/anal_avr.c [hunk 2]
```diff
@@ -17,35 +17,274 @@
...
+	int pc_bits;
+	int pc_mask;
+	int pc_size;
+	int eeprom_size;
+} CPU_MODEL;
+
+typedef void (*inst_handler_t) (RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *buf, int len, int *fail, CPU_MODEL *cpu);
+
+typedef struct _opcodes_tag_ {
+	char *name;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: this is const too

File: libr/anal/p/anal_avr.c [hunk 3]
```diff
@@ -17,35 +17,274 @@
...
+	case 0x3f: /* SREG */ r_strbuf_setf (&op->esil, "r%d,sreg,=", r); break;
+	case 0x3e: /* SPH  */ r_strbuf_setf (&op->esil, "r%d,sph,=",  r); break;
+	case 0x3d: /* SPL  */ r_strbuf_setf (&op->esil, "r%d,spl,=",  r); break;
+	default:
+		r_strbuf_setf (&op->esil, "2,$");
+	}
+}
+
+INST_HANDLER (rcall) {
+	ut32 offset;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: define and initialize in the same line

File: libr/anal/p/anal_avr.c [hunk 4]
```diff
@@ -17,35 +17,274 @@
...
+
+	// target offset
+	offset = ((((buf[1] & 0xf) << 8) | buf[0]) + 1) << 1;
+	if (offset & 0x100)
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: add braces

File: libr/anal/p/anal_avr.c [hunk 5]
```diff
@@ -17,35 +17,274 @@
...
+	op->jump = op->addr + offset;
+	op->fail = 0;
+
+	// cycles!
+	if (!strncasecmp (anal->cpu, "ATtiny", 6)) {
+		op->cycles = 4;	// ATtiny is always slow
+	} else {
+		// PC size decides required runtime!
+		op->cycles = cpu->pc_bits <= 16 ? 3 : 4;
+		if (strncasecmp (anal->cpu, "ATxmega", 7))
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: add braces... and you will probably want to ! this condition

File: libr/anal/p/anal_avr.c [hunk 6]
```diff
@@ -17,35 +17,274 @@
...
+	// op->family = R_ANAL_OP_FAMILY_PRIV;
+
+	// RETI: The I-bit is cleared by hardware after an interrupt
+	// has occurred, and is set by the RETI instruction to enable
+	// subsequent interrupts
+	r_strbuf_append (&op->esil, ",1,if,=");
+}
+
+INST_HANDLER (rjmp) {
+	ut32 offset;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: define and assign in the same line

File: libr/anal/p/anal_avr.c [hunk 7]
```diff
@@ -17,35 +17,274 @@
...
+
+	op->fail = 0;
+
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: unnecessary empty lines

File: libr/anal/p/anal_avr.c (hunk 1 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: moved to const char const *model, which is pretty const

File: libr/anal/p/anal_avr.c (hunk 3 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: fixed

File: libr/anal/p/anal_avr.c (hunk 4 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: fixed

File: libr/anal/p/anal_avr.c (hunk 5 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: yep! true

File: libr/anal/p/anal_avr.c (hunk 7 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: fixed

**Comment by Maijin** [comment]
@killabytenow  https://s3.amazonaws.com/archive.travis-ci.org/jobs/160403786/log.txt search for [XX] to see regressions.

**Comment by radare** [comment]
ping

**Comment by radare** [comment]
please rebase on top of master now that it's green again

**Comment by radare** [comment]
if you dont rebase you will never see travis green

**Comment by Maijin** [comment]
Well the [XX] changed.

**Comment by Maijin** [comment]
Looks almost fixed.

**Comment by radare** [comment]
still 5 broken avr tests

**Pull request merged**